x/distribution/keeper/allocation.go

```diff
@@ -1,57 +1,85 @@
+	// temporary workaround to keep CanWithdrawInvariant happy
+	// general discussions here: https://github.com/cosmos/cosmos-sdk/issues/2906#issuecomment-441867634
+	if totalPower == 0 {
+		feePool.CommunityPool = feePool.CommunityPool.Plus(feesCollected)
 		k.SetFeePool(ctx, feePool)
-		k.feeCollectionKeeper.ClearCollectedFees(ctx)
 		return
 	}
 
-	// allocated rewards to proposer
+	// calculate fraction votes
+	fractionVotes := sdk.NewDec(sumPrecommitPower).Quo(sdk.NewDec(totalPower))
+
+	// calculate proposer reward
 	baseProposerReward := k.GetBaseProposerReward(ctx)
 	bonusProposerReward := k.GetBonusProposerReward(ctx)
-	proposerMultiplier := baseProposerReward.Add(bonusProposerReward.Mul(percentVotes))
-	proposerReward := feesCollectedDec.MulDec(proposerMultiplier)
+	proposerMultiplier := baseProposerReward.Add(bonusProposerReward.Mul(fractionVotes))
+	proposerReward := feesCollected.MulDec(proposerMultiplier)
 
-	// apply commission
-	commission := proposerReward.MulDec(proposerValidator.GetCommission())
-	remaining := proposerReward.Minus(commission)
-	proposerDist.ValCommission = proposerDist.ValCommission.Plus(commission)
-	proposerDist.DelPool = proposerDist.DelPool.Plus(remaining)
+	// pay proposer
+	proposerValidator := k.stakeKeeper.ValidatorByConsAddr(ctx, proposer)
+	k.AllocateTokensToValidator(ctx, proposerValidator, proposerReward)
+	remaining := feesCollected.Minus(proposerReward)
 
-	// allocate community funding
+	// calculate fraction allocated to validators
 	communityTax := k.GetCommunityTax(ctx)
-	communityFunding := feesCollectedDec.MulDec(communityTax)
-	feePool.CommunityPool = feePool.CommunityPool.Plus(communityFunding)
+	voteMultiplier := sdk.OneDec().Sub(proposerMultiplier).Sub(communityTax)
 
-	// set the global pool within the distribution module
-	poolReceived := feesCollectedDec.Minus(proposerReward).Minus(communityFunding)
-	feePool.ValPool = feePool.ValPool.Plus(poolReceived)
+	// allocate tokens proportionally to voting power
+	// TODO consider parallelizing later, ref https://github.com/cosmos/cosmos-sdk/pull/3099#discussion_r246276376
+	for _, vote := range votes {
+		validator := k.stakeKeeper.ValidatorByConsAddr(ctx, vote.Validator.Address)
 
-	k.SetValidatorDistInfo(ctx, proposerDist)
+		// TODO likely we should only reward validators who actually signed the block.
+		// ref https://github.com/cosmos/cosmos-sdk/issues/2525#issuecomment-430838701
+		powerFraction := sdk.NewDec(vote.Validator.Power).Quo(sdk.NewDec(totalPower))
+		reward := feesCollected.MulDec(voteMultiplier).MulDec(powerFraction)
+		k.AllocateTokensToValidator(ctx, validator, reward)
+		remaining = remaining.Minus(reward)
+	}
+
+	// allocate community funding
+	feePool.CommunityPool = feePool.CommunityPool.Plus(remaining)
 	k.SetFeePool(ctx, feePool)
 
-	// clear the now distributed fees
-	k.feeCollectionKeeper.ClearCollectedFees(ctx)
+	// update outstanding rewards
+	outstanding := k.GetOutstandingRewards(ctx)
```

Comment:
Is this just for debugging? If so can we note this somewhere, and perhaps mark that it is for debugging in the variable name?

Reply:
It's useful to track, not very expensive (will always be cached), and allows us a cheap invariant to ensure no tokens are generated.

Reply:
I think we should keep this for the moment, marking "resolved" for now.